src/stores/DAOEntityListStore.js

```diff
@@ -0,0 +1,114 @@
+// @flow
+
+import type { DAO, QueryOptions } from 'brewskey.js-api';
+
+import { action, observable } from 'mobx';
+import flattenArray from 'array-flatten';
+import DAOApi, { LoadObject } from 'brewskey.js-api';
+import { createRange } from '../utils';
+
+export type Row<TEntity> = {|
+  key: number,
+  value: LoadObject<TEntity>,
+|};
+
+class DAOEntityListStore<TEntity> {
+  _baseQueryOptions: QueryOptions = {};
+  _dao: DAO<TEntity, *>;
+  _daoSubscriptionID: string;
+  _pageSize: number;
+
+  @observable _queryOptionsList: Array<QueryOptions> = [];
+
+  @observable rows: Array<Row<TEntity>> = [];
+
+  constructor(
+    dao: DAO<TEntity, *>,
+    queryOptions?: QueryOptions = {},
+    pageSize?: number = 20,
+  ) {
+    this._baseQueryOptions = queryOptions;
+    this._dao = dao;
+    this._pageSize = pageSize;
+
+    this._daoSubscriptionID = this._dao.subscribe(this._computeRows);
+    this.reset();
+  }
+
+  dispose = () => {
+    this._dao.unsubscribe(this._daoSubscriptionID);
+  };
+
+  @action
+  fetchNextPage = () => {
+    const currentQuery = this._queryOptionsList[
+      this._queryOptionsList.length - 1
+    ];
+
+    // prevent fetch when we reached end of the list
+    const currentQueryLoadObject = this._dao.fetchMany(currentQuery);
+    if (
+      currentQueryLoadObject.hasValue() &&
+      currentQueryLoadObject.getValueEnforcing().length < this._pageSize
+    ) {
+      return;
+    }
+
+    this._queryOptionsList.push({
+      ...currentQuery,
+      skip: currentQuery.skip + this._pageSize,
+    });
+    this._computeRows();
+  };
+
+  @action
+  reset = () => {
+    // todo when we use it with onRefresh in FlatList
+    // it works normal but doesn't fit standard mobile refresh behavior.
+    // instead of flushing all the cache we need somehow refetch only
+    // the first query and shift all other cache if there are new entitites,
+    // but it seems too complex.
+    this._dao.flushCache();
+    this._queryOptionsList = [];
+    this._queryOptionsList.push({
+      ...this._baseQueryOptions,
+      skip: 0,
+      take: this._pageSize,
+    });
+    this._computeRows();
+  };
+
+  setQueryOptions = (queryOptions: QueryOptions) => {
+    this._baseQueryOptions = queryOptions;
+    this.reset();
+  };
+
+  @action
+  _computeRows = () => {
+    this.rows = flattenArray(
+      this._queryOptionsList.map((queryOptions: QueryOptions) => {
+        const rowKeys = createRange(
+          queryOptions.skip || 0,
+          queryOptions.take + queryOptions.skip - 1,
+        );
+
+        const queryLoadObject = DAOApi.LocationDAO.fetchMany(queryOptions);
```

Comment:
This can be simplified:
```
const totalEntityCount = this.totalEntityCountLoader.getValueEnforcing();
const itemsToLoadArray = Array.apply(null, Array(totalEntityCount % queryOptions.take || queryOptions.take))
const queryLoadObject = DAOApi.LocationDAO.fetchMany(queryOptions);
if (!queryLoadObject.isLoaded() || queryLoadObject.hasError()) {
  return itemsToLoadArray .map(
    (_, index) => ({
      key: index + queryOptions.skip, 
      value: queryLoadObject,
    }),
  );
}

return queryLoadObject.getValueEnforcing();
```

Reply:
not sure if this is simplification xD, and the logic seems incorrect, in both cases we need to return an array with key/value, but you return just values in the second return case.